{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "from utils import get_sequences\n",
    "from itertools import islice"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# df = pd.read_csv('data/cagi4_mpra/4-eQTL-causal_SNPs_sample_v2.txt', sep='\\t')\n",
    "df = pd.read_csv('data/cagi4_mpra/train_df.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# if the refallele is one character long, it should be 75[1][74]    1 0\n",
    "# if the refallele is two characters long, it should be 74[2][74]   2 1\n",
    "# if the refallele is three characters long, it should be 74[3]73   3 1\n",
    "# if it's four characters long, it should be 73[4][73]              4 2\n",
    "\n",
    "# start_ind is in fact 76 - ((len(RefAllele)+1)// 2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>Unnamed: 0.1</th>\n",
       "      <th>Unnamed: 0.1.1</th>\n",
       "      <th>ID</th>\n",
       "      <th>chr</th>\n",
       "      <th>pos</th>\n",
       "      <th>RefAllele</th>\n",
       "      <th>AltAllele</th>\n",
       "      <th>C.A.ctrl.mean</th>\n",
       "      <th>C.A.exp.mean</th>\n",
       "      <th>...</th>\n",
       "      <th>Regulatory_Hit</th>\n",
       "      <th>emVar_Hit</th>\n",
       "      <th>eQTL_lead</th>\n",
       "      <th>eQTL_gene</th>\n",
       "      <th>eQTL_beta</th>\n",
       "      <th>eQTL_tStat</th>\n",
       "      <th>eQTL_P</th>\n",
       "      <th>sequence</th>\n",
       "      <th>ref_sequence</th>\n",
       "      <th>alt_sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>chr1:183160326:I</td>\n",
       "      <td>1</td>\n",
       "      <td>183160326</td>\n",
       "      <td>G</td>\n",
       "      <td>GAT</td>\n",
       "      <td>169.349792</td>\n",
       "      <td>142.237956</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>chr1:183160326:I</td>\n",
       "      <td>ENSG00000058085</td>\n",
       "      <td>0.128792</td>\n",
       "      <td>6.503906</td>\n",
       "      <td>2.660000e-10</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CAAACAATATTTTTACAATAAAGTAATTGAGAAAGTACCCTATAAT...</td>\n",
       "      <td>CAAACAATATTTTTACAATAAAGTAATTGAGAAAGTACCCTATAAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1:218193345:D</td>\n",
       "      <td>1</td>\n",
       "      <td>218193345</td>\n",
       "      <td>TC</td>\n",
       "      <td>T</td>\n",
       "      <td>546.915129</td>\n",
       "      <td>510.274157</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>chr1:218193345:D</td>\n",
       "      <td>ENSG00000223375</td>\n",
       "      <td>0.181454</td>\n",
       "      <td>7.312115</td>\n",
       "      <td>1.760000e-12</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GGCTTCACACTGTTTTTATTAGGGCTTATTGTTTGGAAAATTAAGT...</td>\n",
       "      <td>GGCTTCACACTGTTTTTATTAGGGCTTATTGTTTGGAAAATTAAGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>chr1:226329182:I</td>\n",
       "      <td>1</td>\n",
       "      <td>226329182</td>\n",
       "      <td>T</td>\n",
       "      <td>TA</td>\n",
       "      <td>117.934774</td>\n",
       "      <td>96.640336</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>chr1:226329182:I</td>\n",
       "      <td>ENSG00000182827</td>\n",
       "      <td>-0.081298</td>\n",
       "      <td>-6.925068</td>\n",
       "      <td>2.050000e-11</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CCTTCAACTCCTGTGCTTAAGTGATCCTCTGGCCTCTGCCTCCCTA...</td>\n",
       "      <td>CCTTCAACTCCTGTGCTTAAGTGATCCTCTGGCCTCTGCCTCCCTA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>chr1:58999158:D</td>\n",
       "      <td>1</td>\n",
       "      <td>58999158</td>\n",
       "      <td>TA</td>\n",
       "      <td>T</td>\n",
       "      <td>178.517618</td>\n",
       "      <td>167.443939</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>chr1:58999158:D</td>\n",
       "      <td>ENSG00000162600</td>\n",
       "      <td>0.099132</td>\n",
       "      <td>7.687867</td>\n",
       "      <td>1.480000e-13</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GTATTAAAATTATGAATGGGAAGCCACATATACTAAAAAGCTCAGT...</td>\n",
       "      <td>GTATTAAAATTATGAATGGGAAGCCACATATACTAAAAAGCTCAGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>chr12:104337222:D</td>\n",
       "      <td>12</td>\n",
       "      <td>104337222</td>\n",
       "      <td>GTAAT</td>\n",
       "      <td>G</td>\n",
       "      <td>86.668936</td>\n",
       "      <td>121.945143</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>chr12:104337222:D</td>\n",
       "      <td>ENSG00000204954</td>\n",
       "      <td>-0.112888</td>\n",
       "      <td>-7.578777</td>\n",
       "      <td>3.070000e-13</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AAGTTATTTTGTGAACAAATTAAGCTGCAGCTGGTTACTTTGTAAC...</td>\n",
       "      <td>AAGTTATTTTGTGAACAAATTAAGCTGCAGCTGGTTACTTTGTAAC...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 31 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   Unnamed: 0  Unnamed: 0.1  Unnamed: 0.1.1                 ID  chr  \\\n",
       "0           0             0               0   chr1:183160326:I    1   \n",
       "1           1             1               1   chr1:218193345:D    1   \n",
       "2           2             2               2   chr1:226329182:I    1   \n",
       "3           3             3               3    chr1:58999158:D    1   \n",
       "4           4             4               4  chr12:104337222:D   12   \n",
       "\n",
       "         pos RefAllele AltAllele  C.A.ctrl.mean  C.A.exp.mean  \\\n",
       "0  183160326         G       GAT     169.349792    142.237956   \n",
       "1  218193345        TC         T     546.915129    510.274157   \n",
       "2  226329182         T        TA     117.934774     96.640336   \n",
       "3   58999158        TA         T     178.517618    167.443939   \n",
       "4  104337222     GTAAT         G      86.668936    121.945143   \n",
       "\n",
       "                         ...                          Regulatory_Hit  \\\n",
       "0                        ...                                       0   \n",
       "1                        ...                                       0   \n",
       "2                        ...                                       0   \n",
       "3                        ...                                       0   \n",
       "4                        ...                                       0   \n",
       "\n",
       "   emVar_Hit          eQTL_lead        eQTL_gene  eQTL_beta  eQTL_tStat  \\\n",
       "0          0   chr1:183160326:I  ENSG00000058085   0.128792    6.503906   \n",
       "1          0   chr1:218193345:D  ENSG00000223375   0.181454    7.312115   \n",
       "2          0   chr1:226329182:I  ENSG00000182827  -0.081298   -6.925068   \n",
       "3          0    chr1:58999158:D  ENSG00000162600   0.099132    7.687867   \n",
       "4          0  chr12:104337222:D  ENSG00000204954  -0.112888   -7.578777   \n",
       "\n",
       "         eQTL_P  sequence                                       ref_sequence  \\\n",
       "0  2.660000e-10       NaN  CAAACAATATTTTTACAATAAAGTAATTGAGAAAGTACCCTATAAT...   \n",
       "1  1.760000e-12       NaN  GGCTTCACACTGTTTTTATTAGGGCTTATTGTTTGGAAAATTAAGT...   \n",
       "2  2.050000e-11       NaN  CCTTCAACTCCTGTGCTTAAGTGATCCTCTGGCCTCTGCCTCCCTA...   \n",
       "3  1.480000e-13       NaN  GTATTAAAATTATGAATGGGAAGCCACATATACTAAAAAGCTCAGT...   \n",
       "4  3.070000e-13       NaN  AAGTTATTTTGTGAACAAATTAAGCTGCAGCTGGTTACTTTGTAAC...   \n",
       "\n",
       "                                        alt_sequence  \n",
       "0  CAAACAATATTTTTACAATAAAGTAATTGAGAAAGTACCCTATAAT...  \n",
       "1  GGCTTCACACTGTTTTTATTAGGGCTTATTGTTTGGAAAATTAAGT...  \n",
       "2  CCTTCAACTCCTGTGCTTAAGTGATCCTCTGGCCTCTGCCTCCCTA...  \n",
       "3  GTATTAAAATTATGAATGGGAAGCCACATATACTAAAAAGCTCAGT...  \n",
       "4  AAGTTATTTTGTGAACAAATTAAGCTGCAGCTGGTTACTTTGTAAC...  \n",
       "\n",
       "[5 rows x 31 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "ref_sequences, alt_sequences = get_sequences(df, which_set='cagi4')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('G', 'C', 'T', 'A')"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# For each variant, a pair of 150-nt candidate regulatory sequence oligonucleotides was synthesized\n",
    "# with the variant located at the central position (i.e. SNP at position 76).\n",
    "sequences[0][75], sequences[1][75], sequences[2][75], sequences[3][75]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "df['ref_sequence'] = ref_sequences\n",
    "df['alt_sequence'] = alt_sequences\n",
    "df.to_csv('data/cagi4_mpra/train_df.csv')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.4.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
