---
title: "Exploration of CAGI5 data"
output:
  html_document:
    theme: united
    highlight: tango
    toc: true
    toc_float:
      collapsed: false
    fig_width: 10
    fig_height: 6
    fig_caption: true
    df_print: paged
    # mathjax: local
    # self_contained: false
---

```{r render, eval = FALSE, echo = FALSE}
#
# Run this chunk to render this document
rmarkdown::render('exploration.Rmd')
```

## Setup
```{r loadPkgs}
# source("https://bioconductor.org/biocLite.R")
# biocLite("Gviz")
suppressPackageStartupMessages({
  library(Gviz)
  data(geneModels)
  library(ggbio)
  data(hg19Ideogram, package = "biovizBase")
  library(GenomicRanges)
  library(BSgenome.Hsapiens.UCSC.hg19)
  library(tidyverse)
  library(reshape2)
  library(stringr)
  library(ggplot2)
  library(ggthemes)
  library(scales)
})
theme_set(theme_tufte())
scale_colour_discrete <- scale_colour_few
scale_fill_discrete <- scale_fill_few
plotdir <- file.path('..', 'plots')
dir.create(plotdir, showWarnings = FALSE)
```


## Regions
```{r regions}
genomeid <- 'hg19'
genomeseqinfo <- Seqinfo(genome = genomeid)

table_str <- 'gene	coords	phenotype	vector	cell_line	Transf_time	fold_change	construct_size
F9	chrX:138612624-138612923	Hemophilia B	pGL4.11b	HepG2	24	2.6	300
GP1BB	chr22:19710790-19711173	Bernard-Soulier Syndrome	pGL4.11b	HEL 92.1.7	24	22.1	384
HBB 	chr11:5248252-5248438	Thalassemia	pGL4.11b	HEL 92.1.7	24	14.3	187
HBG 	chr11:5271035-5271308	Hereditary persistence of fetal hemoglobin	pGL4.11b	HEL 92.1.7	24	118.1	274
HNF4A (P2)	chr20:42984160-42984444	Maturity-onset diabetes of the young (MODY)	pGL4.11b	HEK293T	24	2.8	285
LDLR	chr19:11199907-11200224 	Familial hypercholesterolemia	pGL4.11b	HepG2	24	110.7	318
MSMB	chr10:51548988-51549578	Prostate cancer	pGL4.11b	HEK293T	24	8.4	591
PKLR	chr1:155271187-155271655	Pyruvate kinase deficiency	pGL4.11b	K562	48	29.4	469
TERT	chr5:1295105-1295362	Various types of cancer	pGL4.11b	HEK293T, GBM	24	231.8	258
IRF4	chr6:396143-396593 	Human pigmentation	pGL4.23	SK-MEL-28	24	44.5	451
IRF6	chr1:209989135-209989734 	Cleft lip	pGL4.23	HaCaT	24	17.0	600
MYC	chr8:128413074-128413673 	Various types of cancer (rs6983267)	pGL4.23	HEK293T	32, 20nM LiCl added after 24hr	0.8	600
SORT1	chr1:109817274-109817873	Plasma low-density lipoprotein cholesterol & myocardial infraction	pGL4.23	HepG2	24	235.3	600
ZFAND3 	chr6:37775276-37775853 	Type 2 diabetes	pGL4.23	MIN6	24	14.3	578'
regions <- readr::read_tsv(table_str)
region_gr <- lapply(regions$coords, function(coord) GRanges(coord, seqinfo = genomeseqinfo))
names(region_gr) <- regions$gene

```



## Test data
```{r loadTraining}
training_file = '../data/cagi5_bkpt.csv'
training_df <-
  read_csv(training_file) %>%
  rename("chrom" = !!names(.[2])) %>%
  mutate(
      chrom = str_c('chr', chrom),
      elem = str_sub(regulatory_element, 9))
elements <- unique(training_df$elem)
element_dfs <- training_df %>% split(.$elem)
nrow(training_df)
sample_n(training_df, 50)
```

```{r testGRanges}
element_grs <-
  lapply(
    element_dfs,
    partial(
      makeGRangesFromDataFrame,
      keep.extra.columns = TRUE,
      start.field = 'Pos',
      end.field = 'Pos',
      ignore.strand = TRUE))
```

Plot the training data for each regulatory element.
```{r vizTraining}
plot_element <- function(.df) {

  ggplot(.df, aes(x = Pos, y = Value, colour = Confidence)) +
    geom_hline(yintercept = 0, alpha = .3) +
    geom_point() +
    # geom_smooth(aes(y = abs(Value))) +
    scale_colour_gradient(
        limits = c(0, 1), low = muted('blue'), high = muted('red'))

}
plot_element_gviz <- function(gr) {
  chr <- as.character(chrom(gr)[1])
  genome_tr <- GenomeAxisTrack(chromosome = chr)
  value_tr <- DataTrack(gr, data = 'Value', name = 'Value', col = 'red')
  conf_tr <- DataTrack(gr, data = 'Confidence', name = 'Confidence', col = 'blue')
  gene_tr <- GeneRegionTrack(geneModels, genome = 'hg19', chromosome = chr, name = "genes")
  seq_tr <- SequenceTrack(Hsapiens, chromosome = chr, name = 'sequence')
  plotTracks(
      list(genome_tr, gene_tr, value_tr, seq_tr, conf_tr),
      from = min(start(gr)),
      to = max(end(gr)))
}
for (elem in elements) {
  print(elem)
  print(element_dfs[[elem]]$chrom[1])
  print(plot_element(element_dfs[[elem]]))
  ggsave(
      file.path(plotdir, str_c('value-conf-', elem, '.pdf')),
      width = 8, height = 6, units = 'in')
  # plot_element_gviz(element_grs[[elem]]))
}
```



## Value distribution
```{r valueDist}
ggplot(training_df, aes(x = Value)) + geom_histogram(binwidth=.1)
# ggplot(training_df, aes(x = log10(abs(Value + 0.08)))) + geom_histogram(binwidth=.1)
```
When the distribution of the effect size (Value) is conditioned on the direction we see
an asymmetry. Remember if the confidence is less than .1, there is no significant effect,
i.e. direction = 0.
```{r valueByDirection}
ggplot(training_df, aes(x = Value)) + geom_histogram(binwidth=.1) + facet_wrap(~ class)
ggplot(training_df %>% filter(0 != class), aes(x = abs(Value), fill = factor(class))) +
  geom_histogram(binwidth=.1, position = 'dodge')
```




## Confidence distribution
```{r confDist}
ggplot(training_df, aes(x = Confidence)) + geom_histogram()
```



## Confidence against value
We plot the confidence against the value. Some high absolute values have low confidences and vice versa.
```{r confValue}
ggplot(training_df, aes(x = Value, y = Confidence, colour = elem)) +
  geom_point(alpha = .3) +
  scale_colour_pander()
ggsave(
    file.path(plotdir, 'conf-value-scatter.pdf'),
    width = 8, height = 6, units = 'in')
```



## Spatial effects
Look at whether high values are spatially correlated. We first summarise the
data by position, calculating the maximum absolute value at each position
across all alternate alleles. Then for each pair of positions closer than some
threshold, we threshold the first position has having a high or low value and
investigate if the distributions of values at nearby positions are different
conditioned on this.
```{r effectCorr, fig.width = 16, fig.height = 12}
conf_threshold <- .15
value_threshold <- .3
dist_threshold <- 40
spatial_corr <- function(.df) {
  summary_df <-
    .df %>%
    group_by(Pos) %>%
    summarise(MaxAbsValue = max(abs(Value)), MaxConf=max(Confidence))
  .df1 <-
    summary_df %>%
    rename(Pos1 = Pos, MaxAbsValue1 = MaxAbsValue, MaxConf1 = MaxConf)
  .df2 <-
    summary_df %>%
    rename(Pos2 = Pos, MaxAbsValue2 = MaxAbsValue, MaxConf2 = MaxConf)
  expand.grid(Pos1 = .df1$Pos1, Pos2 = .df2$Pos2) %>%
    mutate(pos_dist = abs(Pos1 - Pos2)) %>%
    filter(pos_dist > 0, pos_dist <= dist_threshold) %>%
    mutate(
        dist_factor =
          cut(pos_dist,
              breaks = seq(0, dist_threshold, by = 5),
              include.lowest = TRUE)) %>%
    left_join(.df1, by = 'Pos1') %>%
    mutate(HighValue1 = MaxAbsValue1 >= value_threshold) %>%
    left_join(.df2, by = 'Pos2') %>%
    mutate(val_diff = abs(MaxAbsValue1 - MaxAbsValue2))
}
spatial_df <- bind_rows(lapply(element_dfs, spatial_corr), .id = 'elem')
ggplot(
    spatial_df,
    aes(x = dist_factor, y = MaxAbsValue2, fill = HighValue1)) +
  geom_boxplot() +
  facet_wrap(~ elem)
ggsave(
    file.path(plotdir, 'spatial.pdf'),
    width = 16, height = 12, units = 'in')
```



## Substitution effects
Look at whether specific substitutions are more likely to have large effect
sizes.
```{r substitutions}

substitutions <-
  training_df %>%
  group_by(elem, Ref, Alt) %>%
  summarise(mean = mean(abs(Value)), sd = sd(abs(Value))) %>%
  mutate(subst = paste(Ref, Alt, sep = ''))
for (e in unique(training_df$elem)) {
  print(e)
  gp <- ggplot(
      training_df %>%
        filter(elem == e) %>%
        mutate(subst = paste(Ref, Alt, sep = '')),
      aes(x = subst, y = Value, fill=Alt)) +
    geom_boxplot(alpha = .5, outlier.alpha = 0) +
    geom_jitter(position=position_jitter(0.2)) +
    facet_grid(~ Ref, scales = 'free_x')
  print(gp)
}

```



## Examine regions in UCSC

```{r examineUCSC, eval = FALSE}
session <- browserSession("UCSC")

by_pos <-
  training_df %>%
  group_by(chrom, Pos) %>%
  summarise(score = max(abs(Value)), MaxConf=max(Confidence))
max_abs_values <-
  makeGRangesFromDataFrame(
      by_pos,
      keep.extra.columns = TRUE,
      ignore.strand = FALSE,
      seqinfo = genomeseqinfo,
      seqnames.field = 'chrom',
      start.field = "Pos",
      end.field = "Pos",
      starts.in.df.are.0based = FALSE)
strand(max_abs_values) <- '+'
max(score(max_abs_values))
score(max_abs_values) <- 1000. * score(max_abs_values) / max(score(max_abs_values))

export(max_abs_values, 'max_abs_values.bigwig', 'BigWig')
track(session, "MaxAbsValues") <- max_abs_values
mcols(track(session, 'MaxAbsValues'))
gr <- region_gr[['PKLR']]
# start(subsetByOverlaps(max_abs_values, gr))
view <- browserView(session, gr, 'MaxAbsValues')

```


## Training data
```{r loadTest}
test_file = '../data/cagi5_mpra/TestDataset.txt'
test_df <- read_tsv(test_file, col_types='ciccc')
nrow(test_df)
sample_n(test_df, 50)
```

