---
title: "Exploration of CAGI5 data"
output:
  html_document:
    theme: united
    highlight: tango
    toc: true
    toc_float:
      collapsed: false
    fig_width: 7
    fig_height: 6
    fig_caption: true
    df_print: paged
    # mathjax: local
    # self_contained: false
---

```{r render, eval = FALSE, echo = FALSE}
rmarkdown::render('exploration.Rmd')
```

## Setup
```{r loadPkgs}
# source("https://bioconductor.org/biocLite.R")
# biocLite("Gviz")
library(Gviz)
library(GenomicRanges)
library(BSgenome.Hsapiens.UCSC.hg19)
library(tidyverse)
library(reshape2)
library(stringr)
library(ggplot2)
library(ggthemes)
data(geneModels)
theme_set(theme_few())
scale_colour_discrete <- scale_colour_few
scale_fill_discrete <- scale_fill_few
```


## Test data
```{r loadTraining}
training_file = '../data/cagi5_bkpt.csv'
training_df <-
  read_csv(training_file) %>%
  rename("chrom" = !!names(.[2])) %>%
  mutate(
      chrom = str_c('chr', chrom),
      elem = str_sub(regulatory_element, 9))
elements <- unique(training_df$elem)
element_dfs <- training_df %>% split(.$elem)
nrow(training_df)
sample_n(training_df, 50)
```

```{r testGRanges}
element_grs <-
  lapply(
    element_dfs,
    partial(
      makeGRangesFromDataFrame,
      keep.extra.columns = TRUE,
      start.field = 'Pos',
      end.field = 'Pos',
      ignore.strand = TRUE))
```

Plot the training data for each regulatory element.
```{r vizTraining}
plot_element <- function(gr) {
  chr <- as.character(chrom(gr)[1])
  genome_tr <- GenomeAxisTrack(chromosome = chr)
  value_tr <- DataTrack(gr, data = 'Value', name = 'Value', col = 'red')
  conf_tr <- DataTrack(gr, data = 'Confidence', name = 'Confidence', col = 'blue')
  gene_tr <- GeneRegionTrack(geneModels, genome = 'hg19', chromosome = chr, name = "genes")
  seq_tr <- SequenceTrack(Hsapiens, chromosome = chr, name = 'sequence')
  plotTracks(
      list(genome_tr, gene_tr, value_tr, seq_tr, conf_tr),
      from = min(start(gr)),
      to = max(end(gr)))
}
for (elem in elements) {
  print(elem)
  print(element_dfs[[elem]]$chrom[1])
  plot_element(element_grs[[elem]])
}
```



## Value distribution
```{r valueDist}
ggplot(training_df, aes(x = Value)) + geom_histogram(binwidth=.1)
```



## Confidence distribution
```{r confDist}
ggplot(training_df, aes(x = Confidence)) + geom_histogram()
```



## Confidence against value
We plot the confidence against the value. Some high absolute values have low confidences and vice versa.
```{r confValue}
ggplot(training_df, aes(x = Value, y = Confidence, colour = elem)) +
  geom_point(alpha = .3) +
  scale_colour_pander()
```



## Spatial effects
Look at whether high values are spatially correlated. We first summarise the
data by position, calculating the maximum absolute value at each position
across all alternate alleles. Then for each pair of positions closer than some
threshold, we threshold the first position has having a high or low value and
investigate if the distributions of values at nearby positions are different
conditioned on this.
```{r effectCorr}
conf_threshold <- .15
value_threshold <- .3
dist_threshold <- 40
spatial_corr <- function(.df) {
  summary_df <-
    .df %>%
    group_by(Pos) %>%
    summarise(MaxAbsValue = max(abs(Value)), MaxConf=max(Confidence))
  .df1 <-
    summary_df %>%
    rename(Pos1 = Pos, MaxAbsValue1 = MaxAbsValue, MaxConf1 = MaxConf)
  .df2 <-
    summary_df %>%
    rename(Pos2 = Pos, MaxAbsValue2 = MaxAbsValue, MaxConf2 = MaxConf)
  expand.grid(Pos1 = .df1$Pos1, Pos2 = .df2$Pos2) %>%
    mutate(pos_dist = abs(Pos1 - Pos2)) %>%
    filter(pos_dist > 0, pos_dist <= dist_threshold) %>%
    mutate(
        dist_factor =
          cut(pos_dist,
              breaks = seq(0, dist_threshold, by = 5),
              include.lowest = TRUE)) %>%
    left_join(.df1, by = 'Pos1') %>%
    mutate(HighValue1 = MaxAbsValue1 >= value_threshold) %>%
    left_join(.df2, by = 'Pos2') %>%
    mutate(val_diff = abs(MaxAbsValue1 - MaxAbsValue2))
}
spatial_df <- bind_rows(lapply(element_dfs, spatial_corr), .id = 'elem')
ggplot(
    spatial_df,
    aes(x = dist_factor, y = MaxAbsValue2, fill = HighValue1)) +
  geom_boxplot() +
  facet_wrap(~ elem)
```




## Training data
```{r loadTest}
test_file = '../data/cagi5_mpra/TestDataset.txt'
test_df <- read_tsv(test_file, col_types='ciccc')
nrow(test_df)
sample_n(test_df, 50)
```

